src/nvim/option.c

```diff
@@ -4888,52 +4900,100 @@ char *set_option_value(const char *const name, const long number,
     return NULL;  // Fail silently; many old vimrcs set t_xx options.
   }
 
-  int opt_idx;
-  char_u      *varp;
+  char *error_message = NULL;
+  int opt_idx = findoption(name);
 
-  opt_idx = findoption(name);
   if (opt_idx < 0) {
     EMSG2(_("E355: Unknown option: %s"), name);
-  } else {
-    uint32_t flags = options[opt_idx].flags;
-    // Disallow changing some options in the sandbox
-    if (sandbox > 0 && (flags & P_SECURE)) {
-      EMSG(_(e_sandbox));
-      return NULL;
+    return NULL;
+  }
+
+  uint32_t flags = options[opt_idx].flags;
+  // Disallow changing some options in the sandbox
+  if (sandbox > 0 && (flags & P_SECURE)) {
+    EMSG(_(e_sandbox));
+    return NULL;
+  }
+
+  if (flags & P_STRING) {
+    const char *s = string;
+    if (s == NULL) {
+      s = "";
     }
-    if (flags & P_STRING) {
-      const char *s = string;
-      if (s == NULL) {
-        s = "";
+    error_message = set_string_option(opt_idx, s, opt_flags);
+    goto after_set;
+  }
+
+  char_u *varp = get_varp_scope(&(options[opt_idx]), opt_flags);
+  if (varp != NULL) {       // hidden option is not changed
+    if (number == 0 && string != NULL) {
+      int idx;
+
+      // Either we are given a string or we are setting option
+      // to zero.
+      for (idx = 0; string[idx] == '0'; idx++) {}
+      if (string[idx] != NUL || idx == 0) {
+        // There's another character after zeros or the string
+        // is empty.  In both cases, we are trying to set a
+        // num option using a string.
+        EMSG3(_("E521: Number required: &%s = '%s'"),
+              name, string);
+        return NULL;  // do nothing as we hit an error
       }
-      return set_string_option(opt_idx, s, opt_flags);
+    }
+
+    if (flags & P_NUM) {
+      error_message = set_num_option(opt_idx, varp, number, NULL, 0, opt_flags);
     } else {
-      varp = get_varp_scope(&(options[opt_idx]), opt_flags);
-      if (varp != NULL) {       // hidden option is not changed
-        if (number == 0 && string != NULL) {
-          int idx;
-
-          // Either we are given a string or we are setting option
-          // to zero.
-          for (idx = 0; string[idx] == '0'; idx++) {}
-          if (string[idx] != NUL || idx == 0) {
-            // There's another character after zeros or the string
-            // is empty.  In both cases, we are trying to set a
-            // num option using a string.
-            EMSG3(_("E521: Number required: &%s = '%s'"),
-                  name, string);
-            return NULL;  // do nothing as we hit an error
-          }
+      error_message = set_bool_option(opt_idx, varp, (int)number, opt_flags);
+    }
+    goto after_set;
+  }
+
+after_set:{};
```

Comment:
Dispatching autocommands right after `set_option_value` (for example, for `'wrap'`) seems to fail at least for `WinScrolled`. My guess is that even though the option has been set, the values that `autocmd_check_window_scrolled` checks to determine if the window has scrolled haven't been changed (values such as `w_left`, `w_top`, etc).

If anyone has hint as to where/when those values are updated after the `set_option_value`, it would be appreciated.